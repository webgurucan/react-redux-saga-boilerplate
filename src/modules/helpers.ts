import { now } from '@gilbarbara/helpers';
import produce from 'immer';
import { Reducer } from 'redux';

import { ActionCreator, ActionsMapReducer, GenericFunction, StoreAction } from 'types';

/**
 * Create an action
 */
export function createAction<T extends GenericFunction>(
  type: string,
  payloadCreator: T,
): ActionCreator<Parameters<T>, ReturnType<T>> {
  if (!payloadCreator) {
    throw new TypeError('Expected a function');
  }

  return (...args: any[]) => ({
    type,
    payload: payloadCreator(...args),
  });
}

/**
 * Create a reducer
 */
export function createReducer<State>(
  actionsMap: ActionsMapReducer<State>,
  defaultState: State,
): Reducer<State, StoreAction> {
  return (state = defaultState, action: StoreAction) =>
    produce(state, (draft: State) => {
      const fn = actionsMap[action.type];

      if (fn) {
        return fn(draft, action);
      }

      return draft;
    });
}

/**
 * Check if cache is valid
 */
export function hasValidCache(lastUpdated: number, max = 10): boolean {
  if (!navigator.onLine) {
    return true;
  }

  return lastUpdated + max * 60 > now();
}
