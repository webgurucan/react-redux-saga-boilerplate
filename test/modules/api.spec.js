import 'isomorphic-fetch';
import nock from 'nock';
import { request } from 'modules/api';

describe('api', () => {
  it('should fail without param', () => {
    expect(() => request()).toThrow('Error! You must pass `url`');
  });

  it('should fail to a POST without payload', () => {
    expect(() => request('http://example.com/token', { method: 'POST' }))
      .toThrow('Error! You must pass `payload`');
  });

  it('should execute a GET successfully', done => {
    nock('http://example.com')
      .get('/token')
      .reply(200, { hello: 'world' });

    request('http://example.com/token')
      .then(data => {
        expect(data).toMatchSnapshot();
        done();
      });
  });

  it('should reject for a server error with JSON response', done => {
    nock('http://example.com')
      .get('/token')
      .reply(500, { error: 'FAILED' });

    request('http://example.com/token')
      .catch(error => {
        expect(error.response).toMatchSnapshot();
        done();
      });
  });

  it('should reject for a server error with no response', done => {
    nock('http://example.com')
      .get('/token')
      .reply(500);

    request('http://example.com/token')
      .catch(error => {
        expect(error.response).toMatchSnapshot();
        done();
      });
  });

  it('should reject for a not found error', done => {
    nock('http://example.com')
      .get('/token')
      .reply(404, {
        error: 'FAILED',
      });

    request('http://example.com/token')
      .catch(error => {
        expect(error.response).toMatchSnapshot();
        done();
      });
  });

  it('should execute a POST successfully', done => {
    nock('http://example.com')
      .post('/token')
      .reply(201);

    request('http://example.com/token', { method: 'POST', payload: { a: 1 } })
      .then(data => {
        expect(data).toMatchSnapshot();
        done();
      });
  });
});
