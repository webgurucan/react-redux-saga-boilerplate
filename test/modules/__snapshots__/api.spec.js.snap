// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api should execute a GET successfully 1`] = `
Object {
  "hello": "world",
}
`;

exports[`api should execute a POST successfully 1`] = `""`;

exports[`api should reject for a not found error 1`] = `
Object {
  "data": Object {
    "error": "FAILED",
  },
  "status": 404,
}
`;

exports[`api should reject for a server error with JSON response 1`] = `
Object {
  "data": Object {
    "error": "FAILED",
  },
  "status": 500,
}
`;

exports[`api should reject for a server error with no response 1`] = `
Object {
  "data": "",
  "status": 500,
}
`;
